### ISSUE TITLE
SPDK initiator does not seem to honor Responder Resource value exchanged during RDMA connection establishment

### ISSUE THREAD

**Issue by sekar-wdc** [spdk/spdk#2261]
Bug report

SPDK initiator does not seem to honor Responder Resource value exchanged during RDMA connection establishment

**Summary of Expected vs Current Behavior**
Details:
During the MAD Connect sequence for establishing an RDMA IO connection for RoCEv2
(NVMe-over Fabrics), the SPDK initiator provides the following values in the REQ message:
• Initiator Depth: 0
• Responder Resources: 16

The target sends a Connect Reply with the following parameters in the REP message:
• Initiator Depth: 16
• Responder Resources: 16

This should result in the target being able to have 16 outstanding RDMA Read Requests at
any time on that IO connection. Based on Ethernet trace analysis, it is observed that the
SPDK initiator sends a NAK (Invalid Request) on reception of the 9th RDMA Read Request. If
the target is modified to ignore the parameters exchanged in the MAD sequence and hard
code the Responder Resources to 8, no issues are observed. What is causing the SPDK
initiator to expect the value of 8 for Responder Resources?

## Context (Environment including OS version, SPDK version, etc.)
CentOS 8.1, SPDK 22.01/21.10/21.07

**Comment by adamagniv** [comment]
any updates?
@sekar-wdc If possible, please also attach your traces.

**Comment by AlekseyMarchuk** [comment]
Hi @adamagniv sorry didn't have time to look at this issue. I plan to check it this week, will update you with my findings

**Comment by sekar-wdc** [comment]
Attached is the trace
[MAT-4700_20211022_1150.zip](https://github.com/spdk/spdk/files/7663648/MAT-4700_20211022_1150.zip)
.

**Comment by AlekseyMarchuk** [comment]
Hi,

I looked at the trace. I found something strange here
First of all, `responder resources` parameter of rdma cm determines the number of RDMA_READ operations that the local side will be ready to accept from the remote side. In NVME-oF protocol only target sends RDMA_READ requests, it never expects RDMA_READ request to be send by the initiator. That means that NVME-oF RDMA target should set `responder_resources=0` and should set `initiator_depth` - this parameter will tell the remote side (initiator) how many RDMA_READ operation the taget will send.  That is done by SPDK, so I think you are using your own target.
On the initiator side these parameters are reversed - it should set `responder_resources` to some value and `initiator_depth=0`.
Please also specify which RDMA NIC you are using and query its `max_qp_rd_atom` and `max_qp_init_rd_atom` parameters. Initiator must not set `responder_resources` more than  `max_qp_rd_atom` and target must not set `initiator_depth` more than `max_qp_init_rd_atom`.
In RDMA operation we have requester and responder, to make it clear NVME-oF RDMA target is the requester of RDMA operation and NVME-oF initiator is the responder
Also in your traces I found some AETH with NAK (`infiniband.aeth.syndrome.opcode==3`) - the first is frame number `2750`. It acks PSN 969744. RDMA_READ with this PSN (`infiniband.bth.psn==969744`) is frame number `2719`. So we know local and remote QP numbers (0x008000 on the target and 0x000f92 on the initiator)
This RDMA_READ wants to read memory with virtual address 35184734695424 and rkey 0x0017f7f5. I tried to find the corresponding nvme command with this virtual address, but didn't find any (`nvme.cmd.sgl1.addr==35184734695424`)
After that I tried to check which nvme commands use rkey 0x0017f7f5 used by this RDMA_READ (`nvme.cmd.sgl1.key == 0x0017f7f5`) and found about 60 packets but no one command requests RDMA_READ to virtual address 35184734695424 (0x2000159CF000). That is why initiator sends NAK for RDMA_READ operation. Please check that your NVME-oF RDMA target uses correct addresses

**Comment by AlekseyMarchuk** [comment]
Update: I see that some nvme commands request to read 64KiB so my last assumption that RMDA_READ uses wrong address may be wrong. But please share your NIC model and supported max_qp_rd_atom and max_qp_init_rd_atom parameters

**Comment by sekar-wdc** [comment]
@AlekseyMarchuk 

The NIC model on the initiator side is a Mellanox CX516A 100GbE with the 16.28.2006 FW.

A little clarification on the trace:

The trace shows:
SPDK initiator (HCA) values in REQ message:
               Initiator Depth:                 0
               Responder Resources:    16
Target (TCA) values in REP message:
               Initiator Depth:                 16
               Responder Resources:    16 (we agree that the target does not expect READ_REQUESTS from initiator and should set this to 0)

The question is that the initiator indicates it can support a maximum of 16 outstanding read Requests and the Target indicates that it will send a maximum of 16 Read Requests per connection; yet, on the 9th outstanding Read Request sent by the target, the initiator sends a NAK. If we change the target to force the local Initiator Depth to 8 (thereby limiting the maximum outstanding Read Request to 8), things work fine at the reduced resource configuration. 

The target does not support Atomic operations.

**Comment by adamagniv** [comment]
@AlekseyMarchuk @sekar-wdc any update?

**Comment by AlekseyMarchuk** [comment]
Yeah, I was looking at your pcap traces at the same moment. I believe the problem is in implementation of RDMA_READ in your target. It looks like the target splits big write nvme commands into several parts and performs several RDMA_READ operations per each nvme command, it may calculate outstanding RDMA_READs incorrectly and limit "responder resources" on the initiator.
That is part of nvme write command sent by SPDK initiator and how SPDK target reads the memory:
```
SGL1
    0100 .... = Descriptor Type: 0x4 Keyed Data Block
    .... 0000 = Descriptor Sub Type: 0x0 Address
    Address: 0x000020000035ac00
    Length: 65536
    Key: 0x0002e7c8
```
And this is part of the corresponding RDMA_READ request from SPDK initiator:
```
RETH - RDMA Extended Transport Header
    Virtual Address: 35184375606272
    Remote Key: 190408
    DMA Length: 65536
```
As you can see, this RDMA operation reads the whole 64KiB payload. 


And here is part of nvme write command from your traces (frame 2682):
```
SGL1
    0100 .... = Descriptor Type: 0x4 Keyed Data Block
    .... 0000 = Descriptor Sub Type: 0x0 Address
    Address: 0x00002000159ee000
    Length: 65536
    Key: 0x0017f7f5
```
And that is RDMA_READ operation sent by your target (frame 2702):
```
RETH - RDMA Extended Transport Header
    Virtual Address: 35184734822400
    Remote Key: 1570805
    DMA Length: 4096
```
As you can see RDMA_READ has length 4096 bytes, SPDK target issues 1 RDMA_READ per incoming write IO while your target seems to split it into several parts. If we add 4096 bytes to `Virtual Address: 35184734822400` then we can find the second part of RDMA_READ operation and so on:
```
RETH - RDMA Extended Transport Header
    Virtual Address: 35184734826496
    Remote Key: 1570805
    DMA Length: 4096
```
Could you check how your target issues RDMA_READ operation? If it splits a request into several parts, it should take into account the number of split RDMA_READ operations not to exceed responder resources

**Comment by sekar-wdc** [comment]
@AlekseyMarchuk 

You are correct that our target does issue RDMA READ REQUESTS in chunks of 4K. However, each RDMA READ REQUEST is treated as a unique request even when they are part of the same NVMe Write command. We will not issue an RDMA READ REQUEST if the outstanding number for the connection has reached the limit negotiated during RDMA connection establishment. In this case, the initiator indicates a value of 16 Responder Resources in the MAD REQ operation. However, it appears to not expect the 9th RDMA Read Request from the target. If we artificially limit the target to send a maximum of 8 outstanding RDMA Read Requests, there is no issue.

**Comment by AlekseyMarchuk** [comment]
@sekar-wdc from pcap traces I see that initiator sends NAK when it receives 16th RDMA_READ request. Have you tried linux kernel initiator? How does it work with your target?

**Comment by sekar-wdc** [comment]
@AlekseyMarchuk There is no similar issue with Linux Inbox driver.

**Comment by AlekseyMarchuk** [comment]
Hi @sekar-wdc we noticed that RDMA initiator is not fully spec compliant in some cases. We are preparing patches to align SPDK initiator with the spec, I think these patches can potentially solve this problem. I'll update you once we push them to gerrit

**Comment by fengchunsong** [comment]
> Hi @sekar-wdc we noticed that RDMA initiator is not fully spec compliant in some cases. We are preparing patches to align SPDK initiator with the spec, I think these patches can potentially solve this problem. I'll update you once we push them to gerrit

Is this problem solved now?

**Comment by AlekseyMarchuk** [comment]
@fengchunsong Could you verify the latest SPDK master branch? We merged several changes for nvme rdma

**Comment by sekar-wdc** [comment]
@fengchunsong @AlekseyMarchuk  Please let me know when the patches are available. We had patched the firmware on our target as a workaround for our customer until the SPDK RDMA initiator was fixed.

**Comment by fengchunsong** [comment]
> @fengchunsong Could you verify the latest SPDK master branch? We merged several changes for nvme rdma
I used SPDK 21.10 to test the write process and got the following error:
[2022-02-16 15:06:37.471223] rdma.c:3778:nvmf_rdma_log_wc_status: *ERROR*: Error on CQ 0xffff4c0921f0, (qp state 1 ibv_state 6) request 0x281468000176408, type SEND, opcode 0, status: (12): transport retry counter exceeded rqpair 0x2c1cc180 byte_len 4096
[2022-02-16 15:06:37.480068] rdma.c: 876:nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16
[2022-02-16 15:07:26.550541] rdma.c:3778:nvmf_rdma_log_wc_status: *ERROR*: Error on CQ 0xffff4c0921f0, (qp state 1 ibv_state 6) request 0x281468000781208, type SEND, opcode 128, status: (12): transport retry counter exceeded rqpair 0x2c3925f0 byte_len 64
[2022-02-16 15:07:26.560236] rdma.c: 876:nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16
[2022-02-16 15:16:22.848546] rdma.c:3778:nvmf_rdma_log_wc_status: *ERROR*: Error on CQ 0xffff400921f0, (qp state 1 ibv_state 6) request 0x281469725830168, type SEND, opcode 128, status: (12): transport retry counter exceeded rqpair 0x2c4d7ea0 byte_len 64
[2022-02-16 15:16:22.853763] rdma.c: 876:nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16
[2022-02-16 15:16:23.372697] rdma.c:3778:nvmf_rdma_log_wc_status: *ERROR*: Error on CQ 0xffff440921f0, (qp state 1 ibv_state 6) request 0x281470342030168, type SEND, opcode 128, status: (12): transport retry counter exceeded rqpair 0x2c0387b0 byte_len 64
[2022-02-16 15:16:23.377062] rdma.c: 876:nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16
[2022-02-16 15:16:23.988254] rdma.c:3778:nvmf_rdma_log_wc_status: *ERROR*: Error on CQ 0xffff08096230, (qp state 1 ibv_state 6) request 0x281467876930968, type SEND, opcode 128, status: (12): transport retry counter exceeded rqpair 0x2c292c80 byte_len 64
[2022-02-16 15:16:23.993781] rdma.c: 876:nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16
[2022-02-16 15:16:26.755630] rdma.c:3778:nvmf_rdma_log_wc_status: *ERROR*: Error on CQ 0xffff64096230, (qp state 1 ibv_state 6) request 0x281470465246808, type SEND, opcode 128, status: (12): transport retry counter exceeded rqpair 0x2c0383c0 byte_len 64
[2022-02-16 15:16:26.763765] rdma.c: 876:nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16
[2022-02-16 15:16:27.179198] rdma.c:3778:nvmf_rdma_log_wc_status: *ERROR*: Error on CQ 0xffff300921f0, (qp state 1 ibv_state 6) request 0x281469232899608, type SEND, opcode 128, status: (12): transport retry counter exceeded rqpair 0x2c414300 byte_len 64
[2022-02-16 15:16:27.183755] rdma.c: 876:nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16
[2022-02-16 15:16:30.848289] rdma.c:3778:nvmf_rdma_log_wc_status: *ERROR*: Error on CQ 0xffff54096230, (qp state 1 ibv_state 6) request 0x281469972745048, type SEND, opcode 128, status: (12): transport retry counter exceeded rqpair 0x2c191710 byte_len 64
[2022-02-16 15:16:30.853758] rdma.c: 876:nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16
[2022-02-16 15:16:31.017875] rdma.c:3778:nvmf_rdma_log_wc_status: *ERROR*: Error on CQ 0xffff100921f0, (qp state 1 ibv_state 6) request 0x281468247299288, type SEND, opcode 0, status: (12): transport retry counter exceeded rqpair 0x2c41caa0 byte_len 64
[2022-02-16 15:16:31.023765] rdma.c: 876:nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16
[2022-02-16 15:16:31.604856] rdma.c:3778:nvmf_rdma_log_wc_status: *ERROR*: Error on CQ 0xffff0c0921f0, (qp state 1 ibv_state 6) request 0x281467753877528, type SEND, opcode 128, status: (12): transport retry counter exceeded rqpair 0x2c4c8f40 byte_len 64
[2022-02-16 15:16:31.613757] rdma.c: 876:nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16
[2022-02-16 15:16:32.438036] rdma.c:3778:nvmf_rdma_log_wc_status: *ERROR*: Error on CQ 0xffff64096230, (qp state 1 ibv_state 6) request 0x281470465581208, type SEND, opcode 128, status: (12): transport retry counter exceeded rqpair 0x2c289810 byte_len 64
[2022-02-16 15:16:32.443749] rdma.c: 876:nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16
[2022-02-16 15:16:33.598498] rdma.c:3778:nvmf_rdma_log_wc_status: *ERROR*: Error on CQ 0xffff500921f0, (qp state 1 ibv_state 6) request 0x281470096017688, type SEND, opcode 128, status: (12): transport retry counter exceeded rqpair 0x2c4d72d0 byte_len 64
[2022-02-16 15:16:33.603757] rdma.c: 876:nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16
[2022-02-16 15:16:33.871973] rdma.c:3778:nvmf_rdma_log_wc_status: *ERROR*: Error on CQ 0xffff28096230, (qp state 1 ibv_state 6) request 0x281468986547928, type SEND, opcode 128, status: (12): transport retry counter exceeded rqpair 0x2bf7c990 byte_len 64
[2022-02-16 15:16:33.875848] rdma.c: 876:nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16
[2022-02-16 15:16:34.244637] rdma.c:3778:nvmf_rdma_log_wc_status: *ERROR*: Error on CQ 0xffff6809b3a0, (qp state 1 ibv_state 6) request 0x281470711921688, type SEND, opcode 2, status: (12): transport retry counter exceeded rqpair 0x2c037be0 byte_len 4096
[2022-02-16 15:16:34.253782] rdma.c: 876:nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16. 

Because RDMA_read is also involved, I doubt it is related to this issue. I want to try to see if the patch that integrates this issue can also fix my problem.

**Comment by sekar-wdc** [comment]
@AlekseyMarchuk Have the fixes to be fully spec compliant been cut in to 22.05 ?

**Comment by AlekseyMarchuk** [comment]
@sekar-wdc There were several changes, I guess this one should help with your problem https://github.com/spdk/spdk/commit/834e3c5a0eab5c2b37f0dc5012ed0822f3d3c4a9
Could you retest your scenario on the latest master branch?
